internal/provider/data_source_gitlab_branch_test.go

```diff
@@ -0,0 +1,110 @@
+package provider
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccDataGitlabBranch_basic(t *testing.T) {
+	rInt := acctest.RandInt()
+	branchProtection := testAccGitlabBranchProtectionAttributes{}
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataGitlabBranch(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceGitlabBranch("gitlab_branch.foo", "data.gitlab_branch.foo"),
+					testAccDataSourceGitlabBranchProtection("data.gitlab_branch.foo", branchProtection),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceGitlabBranch(src, n string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+
+		branch := s.RootModule().Resources[src]
+		branchAttr := branch.Primary.Attributes
+
+		search := s.RootModule().Resources[n]
+		searchAttr := search.Primary.Attributes
+
+		testAttributes := []string{
+			"id",
+			"name",
+			"web_url",
+			"default",
+			"project",
+			"can_push",
+			"merged",
+			"commit",
+			"parent_ids",
+		}
+
+		for _, attribute := range testAttributes {
+			if searchAttr[attribute] != branchAttr[attribute] {
+				return fmt.Errorf("expected branch's parameter `%s` to be: %s, but got: `%s`", attribute, branchAttr[attribute], searchAttr[attribute])
+			}
+		}
+		return nil
+	}
+}
+
+func testAccDataSourceGitlabBranchProtection(n string, exp testAccGitlabBranchProtectionAttributes) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		search := s.RootModule().Resources[n].Primary.Attributes
+		if search["protected"] != strconv.FormatBool(exp.Protected) {
+			return fmt.Errorf("expected parameter protected  to be: %s, but got: `%v`", search["protected"], exp.Protected)
+		}
+		if search["developer_can_merge"] != strconv.FormatBool(exp.DevelopersCanMerge) {
+			return fmt.Errorf("expected parameter developer_can_merge  to be: %s, but got: %v", search["developer_can_merge"], exp.DevelopersCanMerge)
+		}
+		if search["developer_can_push"] != strconv.FormatBool(exp.DevelopersCanPush) {
+			return fmt.Errorf("expected parameter developer_can_push  to be: %s, but got: %v", search["developer_can_push"], exp.DevelopersCanPush)
+		}
+		return nil
+	}
+}
+
+func testAccDataGitlabBranch(rInt int) string {
+	return fmt.Sprintf(`
+%s
+
+data "gitlab_branch" "foo" {
+  name = "${gitlab_branch.foo.name}"
+  project = "${gitlab_branch.foo.project}"
+}
+`, testAccDataGitlabBranchSetup(rInt))
+}
+
+func testAccDataGitlabBranchSetup(rInt int) string {
```

Comment:
Let's focus the test on the data source / resource being tested.

It would be nice to use the `testGitlabClient` directly for setup code. See the helper functions in `helper_test.go` for examples.

Reply:
I also assume that depending on the execution order the `data.gitlab_branch.foo` wouldn't find the branch ...

Reply:
> I also assume that depending on the execution order the `data.gitlab_branch.foo` wouldn't find the branch ...

Do you mean we should test the case where it doesnt find a branch?? It wouldnt be able to execute in any other order due to the dependencies in the terraform configuration itself ?

Reply:
Yes, just noticed. Nevermind ;)

Still the first comment applies nonetheless :)

Reply:
Do you mean create the project and group through direct api creation methods using the helper rather than via terraform config in the tests, I just followed how the other tests were written generally -- personally think it's better this way as it's self documenting in how to use the data source / resource in an actual terraform block -- also potentially a better integration test than just assuming it works well with other terraform resources and going direct via api

Reply:
I agree to these upsides, however, this test also breaks if there is a bug in let's say the `gitlab_project` resource, which really is suboptimal. I'd rather want the test itself to fail for as few reasons as possible w.r.t to production code (of course there may be issues in the test code itself, too).

Reply:
--- FAIL: TestAccDataGitlabBranch_basic (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x2d8 pc=0x180a06e]

goroutine 20 [running]:
testing.tRunner.func1.2({0x1953860, 0x22a53f0})
        /usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x1953860, 0x22a53f0})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/gitlabhq/terraform-provider-gitlab/internal/provider.testAccCreateProject(0xc00009b860)
        /Users/Selecta/Documents/atosGO/terraform-provider-gitlab/internal/provider/helper_test.go:103 +0x16e
github.com/gitlabhq/terraform-provider-gitlab/internal/provider.TestAccDataGitlabBranch_basic(0xc00009b860)
        /Users/Selecta/Documents/atosGO/terraform-provider-gitlab/internal/provider/data_source_gitlab_branch_test.go:14 +0x4c
testing.tRunner(0xc00009b860, 0x1b01348)
        /usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1306 +0x35a
FAIL    github.com/gitlabhq/terraform-provider-gitlab/internal/provider 0.293s
FAIL
make: *** [test] Error 1

Reply:
Looks resolved